third_party/terraform/resources/resource_compute_instance_template.go

```diff
@@ -726,46 +727,212 @@ func resourceComputeInstanceTemplateCreate(d *schema.ResourceData, meta interfac
+	}
+
+	if disk.DiskEncryptionKey != nil {
+		encryption := make([]map[string]interface{}, 1)
+		encryption[0] = make(map[string]interface{})
+		encryption[0]["kms_key_self_link"] = disk.DiskEncryptionKey.KmsKeyName
+		diskMap["disk_encryption_key"] = encryption
+	}
+
+	diskMap["auto_delete"] = disk.AutoDelete
+	diskMap["boot"] = disk.Boot
+	diskMap["device_name"] = disk.DeviceName
+	diskMap["interface"] = disk.Interface
+	diskMap["source"] = ConvertSelfLinkToV1(disk.Source)
+	diskMap["mode"] = disk.Mode
+	diskMap["type"] = disk.Type
+
+	return diskMap, nil
+}
+
+func reorderDisks(configDisks []interface{}, apiDisks []map[string]interface{}) []map[string]interface{} {
+	if len(apiDisks) != len(configDisks) {
+		// There are different numbers of disks in state and returned from the API, so it's not
+		// worth trying to reorder them since it'll be a diff anyway.
+		return apiDisks
+	}
+
+	result := make([]map[string]interface{}, len(apiDisks), len(apiDisks))
+
+	/*
+		Disks aren't necessarily returned from the API in the same order they were sent, so gather
+		information about the ones in state that we can use to map it back. We can't do this by
+		just looping over all of the disks, because you could end up matching things in the wrong
+		order. For example, if the config disks contain the following disks:
+		disk 1: auto delete = false, size = 10
+		disk 2: auto delete = false, size = 10, device name = "disk 2"
+		disk 3: type = scratch
+		And the disks returned from the API are:
+		disk a: auto delete = false, size = 10, device name = "disk 2"
+		disk b: auto delete = false, size = 10, device name = "disk 1"
+		disk c: type = scratch
+		Then disk a will match disk 1, disk b won't match any disk, and c will match 3, making the
+		final order a, c, b, which is wrong. To get disk a to match disk 2, we have to go in order
+		of fields most specifically able to identify a disk to least.
+	*/
+	disksByDeviceName := map[string]int{}
+	scratchDisksByInterface := map[string][]int{}
+	attachedDisksBySource := map[string]int{}
+	attachedDisksByDiskName := map[string]int{}
+	attachedDisksByCharacteristics := []int{}
+
+	for i, d := range configDisks {
+		if i == 0 {
+			// boot disk
+			continue
+		}
+		disk := d.(map[string]interface{})
+		if v := disk["device_name"]; v.(string) != "" {
+			disksByDeviceName[v.(string)] = i
+		} else if v := disk["type"]; v.(string) == "SCRATCH" {
+			iface := disk["interface"].(string)
+			if iface == "" {
+				// apply-time default
+				iface = "SCSI"
+			}
+			scratchDisksByInterface[iface] = append(scratchDisksByInterface[iface], i)
+		} else if v := disk["source"]; v.(string) != "" {
+			attachedDisksBySource[v.(string)] = i
+		} else if v := disk["disk_name"]; v.(string) != "" {
+			attachedDisksByDiskName[v.(string)] = i
+		} else {
+			attachedDisksByCharacteristics = append(attachedDisksByCharacteristics, i)
+		}
+	}
+
+	// Align the disks, going from the most specific criteria to the least.
```

Comment:
I know that it won't help for the initial Create, but have you looked at using the [AttachedDisk.Index](https://godoc.org/google.golang.org/api/compute/v1#AttachedDisk)?

Reply:
Yeah, there's no way to set it but we could store it. I think we'd have to do the full logic though for the initial create, in which case may as well do it the rest of the time (you can't actually modify instance templates so I don't expect us to be catching a lot of drift)